### PR TITLE
Resolve #1126: Prepare for cascades planner deployment

### LINKED ISSUE #1126: Cascades planner prep
Preparation for the cascades planner:
- Add properties to protect from planning infinite loop and memory abuse
- Add setCascadesConfiguration() to cascades 
These are needed in order to ensure the planner is not abusing resources.

### PR DESCRIPTION
**Pull request by ohadzeliger**
Resolve #1126

### CODE REVIEW

**Comment by ohadzeliger** [comment]
retest this please

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java [hunk 1]
```diff
@@ -37,19 +37,25 @@
     private final boolean checkForDuplicateConditions;
     private final boolean deferFetchAfterUnionAndIntersection;
     private final boolean optimizeForIndexFilters;
+    private final int maxTaskQueueSize;
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java:40]: Can you add a comment somewhere that these parameters being `0` (which is their default) indicate that we don't enforce the limit?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java [hunk 2]
```diff
@@ -279,6 +302,35 @@ public void setIndexScanPreference(@Nonnull IndexScanPreference indexScanPrefere
...
+    @Nonnull
+    public RecordQueryPlannerConfiguration getConfiguration() {
+        return configuration;
+    }
+
+    public void setConfiguration(@Nonnull final RecordQueryPlannerConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    private boolean taskQueueSizeExceeded(final RecordQueryPlannerConfiguration configuration, final int queueSize) {
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:326]: can you change name to `is.....` (code style).

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java [hunk 3]
```diff
@@ -279,6 +302,35 @@ public void setIndexScanPreference(@Nonnull IndexScanPreference indexScanPrefere
...
+        return ((configuration.getMaxTaskQueueSize() > 0) && (queueSize > configuration.getMaxTaskQueueSize()));
+    }
+
+    private boolean taskTotalCountExceeded(final RecordQueryPlannerConfiguration configuration, final int taskCount) {
```
Comment (normen662) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java:330]: can you change name to is..... (code style).

**Review by normen662 — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java (hunk 1 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlannerConfiguration.java:40]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java (hunk 2 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java (hunk 3 above)
Comment (ohadzeliger) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/CascadesPlanner.java]: Done.

**Review by normen662 — APPROVED** [review]

**Pull request merged**